docs/src/main/sphinx/develop/connectors.rst

```diff
@@ -79,33 +79,103 @@ features, like:
 Note that data modification also requires implementing
 a :ref:`connector-page-sink-provider`.
 
+When a ``SELECT`` query is being executed, and it contains specific clauses,
+the engine can try pushing down some of them into the connector, by calling
+one of the following methods of the ``ConnectorMetadata`` service:
+
+* ``applyLimit`` for the ``LIMIT`` clause, if ``ORDER BY`` is not present,
+  or ``applyTopN`` when ``ORDER BY`` is present
+* ``applyFilter`` for the ``WHERE`` clause
+* ``applyProjection`` for expressions in the ``SELECT`` clause
+* ``applySample`` for the ``TABLESAMPLE`` clause
+* ``applyAggregation`` for the ``GROUP BY`` clause
+* ``applyJoin`` for the different types of the ``JOIN`` clause
+* ``applyTableFunction`` for the ``TABLE ()`` table function clause
+* ``applyTableScanRedirect`` to redirect table scans to a different connector
+
+
+Connectors can indicate that they don't support a particular pushdown or that
+the action had no effect by returning ``Optional.empty()``. Connectors should
+expect these methods to be called multiple times during the optimization of
+a given query.
+
+.. warning::
+
+  It's critical for connectors to return ``Optional.empty()`` if calling
+  this method has no effect for that invocation, even if the connector generally
+  supports a particular pushdown. Doing otherwise can cause the optimizer
+  to loop indefinitely.
+
+Otherwise, these methods return a result object containing the connectors
+table handle. The connector can store any kind of information in the table
+handle, as it will be passed later to other services it implements, like
+the ``ConnectorRecordSetProvider`` or ``ConnectorPageSourceProvider``.
+
+Limit and TopN pushdown
+^^^^^^^^^^^^^^^^^^^^^^^
+
+When a ``SELECT`` query contains a :ref:`limit-clause`, the engine will try to
+push down the limit into the connector. If there's no ``ORDER BY`` clause,
+the ``applyLimit`` method of the connector metadata service will be called,
+and the connector can return results in an arbitrary order.
+When there's a ``ORDER BY`` clause present, the engine will call ``applyTopN``.
```

Comment:
This needs to be described in terms of IR elements, not SQL-level constructs

Reply:
Can you give one example? I can only think about something like `When the plan contains operations related to a ... clause` but it seems too verbose to use too often.

Reply:
E.g.,

> When the plan contains a `Sort` operation followed by a `Limit` operation, the engine will call `applyTopN`

Reply:
Thank you so much. I redacted this section. PTAL as it still might be phrased better, but I now feel more confident in writing about pushdowns.